oxidation/crates/parsec_api_crypto/src/rustcrypto/private.rs

```diff
@@ -130,65 +126,49 @@ impl PrivateKey {
     }
 
     pub fn generate_shared_secret_key(&self, peer_public_key: &PublicKey) -> SecretKey {
-        SecretKey::from(x25519(
-            *self.0.as_bytes(),
-            peer_public_key.0.as_bytes().to_owned(),
-        ))
+        SecretKey::from(x25519(*self.0.as_bytes(), *peer_public_key.0.as_bytes()))
     }
 }
 
 impl AsRef<[u8]> for PrivateKey {
-    #[inline]
     fn as_ref(&self) -> &[u8] {
-        // Unsafe on private key, what could possibly go wrong ?
-        // TODO: crypto_box::SecretKey should implement AsRef<[u8]> instead
-        unsafe {
-            ::std::slice::from_raw_parts(
-                (self as *const Self) as *const u8,
-                ::std::mem::size_of::<Self>(),
-            )
-        }
+        self.0.as_bytes()
     }
 }
 
 impl TryFrom<&[u8]> for PrivateKey {
     type Error = CryptoError;
     fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
-        let key: [u8; KEY_SIZE] = data.try_into().map_err(|_| CryptoError::DataSize)?;
-        Ok(Self(crypto_box::SecretKey::from(key)))
+        <[u8; Self::SIZE]>::try_from(data)
+            .map(Self::from)
+            .map_err(|_| CryptoError::DataSize)
     }
 }
 
 impl From<[u8; Self::SIZE]> for PrivateKey {
     fn from(key: [u8; Self::SIZE]) -> Self {
-        // TODO: zerocopy
-        Self::try_from(key.as_ref()).unwrap()
+        Self(key.into())
     }
 }
 
-impl TryFrom<ByteBuf> for PrivateKey {
+impl TryFrom<&Bytes> for PrivateKey {
     type Error = CryptoError;
-    fn try_from(data: ByteBuf) -> Result<Self, Self::Error> {
-        let key: [u8; KEY_SIZE] = data
-            .to_vec()
-            .try_into()
-            .map_err(|_| CryptoError::DataSize)?;
-        Ok(Self(crypto_box::SecretKey::from(key)))
+    fn try_from(data: &Bytes) -> Result<Self, Self::Error> {
+        Self::try_from(data.as_ref())
     }
 }
 
-impl From<PrivateKey> for ByteBuf {
-    fn from(data: PrivateKey) -> Self {
-        Self::from(*data.0.as_bytes())
+impl Serialize for PrivateKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_bytes(self.as_ref())
     }
 }
 
-/*
```

Comment:
We talk about this, those comments are here to stay.

Reply:
We could make these comments more useful than just being empty

Reply:
Ok

Reply:
@FirelightFlagboy the need is to separate the different things behind implemented in this file, those comments do this pretty well

Now having those comment to more than that is another topic ;-)

Reply:
why not splitting the file then ?